src/Controller/RestApiController.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Www\Controller;
+
+use Www\DataObject\RestResponseDo;
+use Www\Exception\RestException;
+use YapepBase\Config;
+use YapepBase\Controller\HttpController;
+use YapepBase\Exception\ControllerException;
+use YapepBase\Mime\MimeType;
+use YapepBase\Request\IRequest;
+use YapepBase\Response\IResponse;
+use YapepBase\View\RestTemplate;
+
+abstract class RestApiController extends HttpController
+{
+    /** @var RestResponseDo */
+    protected $restResponseDo;
+
+    /** @var string */
+    private $requestBody;
+
+    /** @var array|null */
+    private $requestData;
+
+    public function __construct(IRequest $request, IResponse $response)
+    {
+        parent::__construct($request, $response);
+        $this->restResponseDo = new RestResponseDo();
+    }
+
+    public function run($action)
+    {
+        try {
+            $this->requestBody = file_get_contents('php://input');
+
+            if ($this->requestBody) {
+                $this->requestData = json_decode($this->requestBody, true);
+            }
+
+            // Initialize the content type to JSON
+            $this->response->setContentType(MimeType::JSON);
+
+            $methodName = $this->getActionPrefix() . $action;
+            if (!method_exists($this, $methodName)) {
+                throw new RestException(
+                    RestException::CODE_METHOD_NOT_SUPPORTED,
+                    'The method ' . $this->request->getMethod() . ' is not supported by this endpoint'
+                );
+            }
+
+            parent::run($action);
+        } catch (RestException $e) {
+            $view = new RestTemplate();
+            if ($this->response->getStatusCode() < 400) {
+                $this->response->setStatusCode($e->getDefaultHttpStatusCode());
+            }
+
+            if (401 == $this->response->getStatusCode() && !$this->response->hasHeader('WWW-Authenticate')) {
+                $this->response->setHeader('WWW-Authenticate', 'Session realm="Please provide the session token"');
+            }
+
+            if (405 == $this->response->getStatusCode() && !$this->response->hasHeader('Allow')) {
+                $this->response->setHeader('Allow', implode(', ', $this->getValidMethodsForAction($action)));
+            }
+
+            $view->setContent($this->getErrorResponse($e->getErrorCode(), $e->getMessage(), $e->getParams()));
+            $view->setContentType($this->response->getContentType());
+
+            $this->response->setBody($view);
+        } catch (\Exception $e) {
+            trigger_error(
+                'Unhandled exception of ' . get_class($e) . ' while serving api response. Error: ' . $e->getMessage(),
+                E_USER_ERROR
```

Comment:
If you use a warning instead of error you're safe, though I know it should be an error.
Or you just leave it like this and add a comment to the class documentation that the Error handling should take care of USER_ERROR-s to not to interrupt the script